### PR TITLE
Setting based permissions - downport

### PR DESCRIPTION
**Pull request by mrsimpson**
This is a downport of https://github.com/RocketChat/Rocket.Chat/pull/8942 : The PR towards Rocket.Chat has been implemented based on `develop` of Rocket.Chat which is 0.60-based.
Thus, I cherry-picked all the commits and verified 0.59.3-compatibility. I didn't do any adaptions to that so far.

Thus, I'd like you (dear reviewer) to comment on https://github.com/RocketChat/Rocket.Chat/pull/8942 for two reasons:

- I already explained some aspects here
- Other reviewers @RocketChat can benefit from your comments as well.

# What to do?

- Check the functionality: Log-in with admin and non-admin-user, giving the non-admin the permissions `manage-selected-permissions` and some setting-based permissions.
- Watch the ability of changing dedicated settings change for the non-admin
- Revert the permissions
- Verify no conflicts to the standard `view/change-priviledged-setting`-permission
- Check the Code ;)

### CODE REVIEW

**Comment by ruKurz** [comment]
Tested functionality:
- the permisson `manage-selected-permissions` does not exist, instead it's named `manage-selected-settings`
- By giving a non-admin role this permission, he is able to access the admin view
- When opening the admin view the `Info` option is available
- After permitting the same non-admin role some setting-based permissions, everything works like charm
- I tested by changing the layout colors, nothing else
- changing dedicated settings works as suspected - also in a reactive way, using two browsers (Safari and chrome)
- In addtion I tested `access-setting-permissions` for non-admin users, also this works well. Only if the user hast both `access-permissions` and `access-setting-permissions` he is able to see the permission option within the admin view. I did not test this via API.
- Removing permissions takes effect in place

I did not get how to test:
- Verify no conflicts to the standard view/change-priviledged-setting-permission

The code looks very straight forward. Since most changes are declarative and not logical. It looks good to me.

Well done!

**Comment by ruKurz** [comment]
I just give a look to the travis build failure

```[0m  1) [Permissions] change the permissions: it should change the create c room permission:
[0m[31m     Uncaught Error: An element could not be located on the page using the given search parameters ("[name="perm[user][create-c]"]").[0m[90m
      at node_modules/chimp/dist/lib/utils/fiberize.js:29:22
      at isSelected() - at Context.<anonymous> (tests/end-to-end/ui/13-permissions.js:117:31)
[0m
[0m  2) [Permissions] "after all" hook:
[0m[31m     Uncaught Error: unknown error: Element <div class="sidebar__account active" aria-label="Account" role="region">...</div> is not clickable at point (140, 30). Other element would receive the click: <h1 class="sidebar-flex__title">...</h1>[0m[90m
      at checkIfUserIsAdmin (tests/data/checks.js:75:23)
      at Context.<anonymous> (tests/end-to-end/ui/13-permissions.js:28:3)
      at node_modules/chimp/dist/lib/utils/fiberize.js:29:22
      at new RuntimeError (node_modules/webdriverio/build/lib/utils/ErrorHandler.js:144:12)
      at Request._callback (node_modules/webdriverio/build/lib/utils/RequestHandler.js:313:39)
      at Request.self.callback (node_modules/request/request.js:186:22)
      at Request.<anonymous> (node_modules/request/request.js:1163:10)
      at IncomingMessage.<anonymous> (node_modules/request/request.js:1085:12)
      at endReadableNT (_stream_readable.js:923:12)
[0m


Mocha failed
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! Rocket.Chat@0.59.3 chimp-test: `chimp tests/chimp-config.js`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the Rocket.Chat@0.59.3 chimp-test script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/travis/.meteor/packages/meteor-tool/.1.5.2_2.fkst7p++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/.npm/_logs/2017-11-29T13_26_18_402Z-debug.log
Chimp exited with code 2
npm ERR! Test failed.  See above for more details.
```

Even this error is part of the headless tests, this error is semantically related with `permissions`. Also it does not seem to be caused by the changes of this PR.

File: package.json [hunk 1]
```diff
@@ -60,7 +60,9 @@
 		"chat"
 	],
 	"scripts": {
-		"start": "meteor npm i && meteor",
+		"start": "meteor npm i && meteor run",
+		"debug": "meteor run --inspect",
+		"debug-brk": "meteor run --inspect-brk",
```
Comment (ruKurz) [inline package.json]: Is this related with the subject of this PR? Or just a general extension for easy debugging?

**Comment by vickyokrm** [comment]
Reviewed and tested the function together with @ThomasRoehl 
Here are the review comments.
- Detailed walk through on the source code implementation.
- The configurable permissions settings list are long which is leads to scrolling difficulties.
(Since, this is screen is used not very often, adding a settings search function to the screen is not handled at this point in time but in future)
- Also, the permission setting named together with the group name and hierarchies which will serve the user to drill down to a particular setting easily.
- The functionality expected is well achieved.
- Changing settings is intuitive in reactive way and its blazing. 
- Tested activating/deactivating the individual settings - all permissions worked as expected.
- Tested the 'manage-selected-settings' enabler option which will enable the administration view for the non-admin user.
- As @ruKurz also highlighted, the 'info' option is displayed for the non-admin user in the administration view even though the non-admin user has any setting to view.
